### DISCUSSION TITLE
mod_oauth2 : Directive for additional post params, similar to OIDCOAuthIntrospectionEndpointParams

### DISCUSSION THREAD

**Discussion by ravikaycee** [OpenIDC/mod_oauth2#44] [Q&A] [answered]
Could you please share the similar directive for mod_oauth2 

OIDCOAuthIntrospectionEndpointParams

It will be useful as there is a need to pass few more mandatory params to introspection url apart from client_id and client_secret.

**Comment by zandbelt** [comment, marked as answer]
that was still a TODO in liboauth2; I just added support for it in https://github.com/zmartzone/liboauth2/commit/090e212c95c8e3d6006955c45bc47f61cab80248 ; would you be able to use a self-compiled version of liboauth2 from there?

**Reply by ravikaycee** [reply]
Thanks for the prompt response. I will explore on how to self compile this version..

**Reply by zandbelt** [reply]
never mind, I'm releasing 1.4.5.5 with support for `introspect.params` as we speak; RPMs have been added already, DEBs will soon follow

**Reply by ravikaycee** [reply]
That's really great. Thanks a lot.

**Reply by ravikaycee** [reply]
Facing this issue while installing. Could you please share your suggestions to resolve this,

Unpacking liboauth2-dev (1.4.5.5-1.bullseye) ...
dpkg: dependency problems prevent configuration of liboauth2-dev:
#0 1.832  liboauth2-dev depends on liboauth2 (= 1.4.5.5-1.bullseye); however:
#0 1.832   Package liboauth2 is not installed.

**Reply by zandbelt** [reply]
you'll need liboauth2 rather than liboauth2-dev

**Reply by ravikaycee** [reply]
[Oauth2_log.txt](https://github.com/zmartzone/mod_oauth2/files/10869534/Oauth2_log.txt)

Trying to pass grant_type=client_credentials as additional param in the introspection url. Getting response as grant_type is missing even after using introspection.params. Could you please check the below config and the attached logs and please let me know what needs to be added in the config to include this parameter in the request.

AuthType oauth2		
		
OAuth2TokenVerify introspect https://auth-server/oauth/token introspect.ssl_verify=false&introspect.auth=client_secret_post&client_id=<client_id>&client_secret=<client_secret>&introspect.params=grant_type%3Dclient_credentials%26
OAuth2TargetPass remote_user_claim=scope
Require oauth2_claim  [ valid-user | "scope: read:users" ]

**Reply by zandbelt** [reply]
at least you'll need to get rid of the trailing `%26` but you should also make sure you actually use liboauth2 1.4.5.5 because when I copy your config, I see the parameter is added to the request

also change the Require line to `Require oauth2_claim scope:read:users` and you probably want to set `remote_user_claim` to something that identifies a user

**Reply by ravikaycee** [reply]
The following packages are installed where the latest version of liboauth2 and liboauth2-apache are used. LoadModule oauth2_module is included in config and oauth2 is used. Apart from this, is there anything else should be updated to ensure liboauth2 1.4.5.5 is used..

RUN dpkg -i liboauth2_1.4.5.5-1.bullseye_amd64.deb
RUN dpkg -i liboauth2-apache_1.4.5.5-1.bullseye_amd64.deb
RUN dpkg -i libapache2-mod-oauth2_3.3.1-1.bullseye_amd64.deb
RUN dpkg -i libapache2-mod-auth-openidc_2.4.12.3-1.bullseye_amd64.deb

**Reply by zandbelt** [reply]
you'll also need to restart the server; upon startup if will display versions used in the log

**Reply by ravikaycee** [reply]
After installing 1.5 version of liboauth2 and liboauth2-apache, I could see introspection.params value in the logs in few places mentioned below, but its not added in oauth2_http_call. Could you please let me know what needs to be added to include grant_type in the introspection request.

oauth2_cfg_token_verify_add_options: enter: type=introspect, value=https://dev-e0e5ikul65zavfux.us.auth0.com/oauth/token, options=introspect.ssl_verify=false&introspect.params=grant_type%3Dclient_credentials%26audience%3Dhttps://dev-e0e5ikul65zavfux.us.auth0.com/api/v2/&introspect.auth=client_secret_basic&client_id=2pACQ7dLngWwxyT1PXbJ02ncFStZyD22&client_secret=L8LjMTE8Jl8_LWsATN_GZoc8G6mnzL1ZG-XAQa4qXVhYvC7NrpShcyWuk9pywgQE

oauth2_url_decode: enter: grant_type%3Dclient_credentials%26audience%3Dhttps://dev-e0e5ikul65zavfux.us.auth0.com/api/v2/

oauth2_url_decode: leave: grant_type=client_credentials&audience=https://dev-e0e5ikul65zavfux.us.auth0.com/api/v2/

oauth2_cfg_set_options: enter: type=introspect, value=https://dev-e0e5ikul65zavfux.us.auth0.com/oauth/token, options=introspect.ssl_verify=false&introspect.params=grant_type%3Dclient_credentials%26audience%3Dhttps://dev-e0e5ikul65zavfux.us.auth0.com/api/v2/&introspect.auth=client_secret_basic&client_id=2pACQ7dLngWwxyT1PXbJ02ncFStZyD22&client_secret=L8LjMTE8Jl8_LWsATN_GZoc8G6mnzL1ZG-XAQa4qXVhYvC7NrpShcyWuk9pywgQE

oauth2_http_call: enter: url=https://dev-e0e5ikul65zavfux.us.auth0.com/oauth/token, data=token=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IjJxMWdrYmJad1pMZTB6cXNZTkxSbCJ9.eyJpc3MiOiJodHRwczovL2Rldi1lMGU1aWt1bDY1emF2ZnV4LnVzLmF1dGgwLmNvbS8iLCJzdWIiOiIycEFDUTdkTG5nV3d4eVQxUFhiSjAybmNGU3RaeUQyMkBjbGllbnRzIiwiYXVkIjoiaHR0cHM6Ly9kZXYtZTBlNWlrdWw2NXphdmZ1eC51cy5hdXRoMC5jb20vYXBpL3YyLyIsImlhdCI6MTY3ODM2MjQzNiwiZXhwIjoxNjc4NDQ4ODM2LCJhenAiOiIycEFDUTdkTG5nV3d4eVQxUFhiSjAybmNGU3RaeUQyMiIsInNjb3BlIjoicmVhZDp1c2VycyIsImd0eSI6ImNsaWVudC1jcmVkZW50aWFscyJ9.HcMVBnT_XFduTenoLQd3PyUFTcw0_QJxFC36ma2-JSgmK20gH6KCncpwPtCUFgse1FB9AQzUefVLW2CVGDBt1wDiQQprD6r06U1fKqfQlWD8bRJw5RS1vlixNlszAb1t7uX1pSOndX-4SAxe8EGe6_EvkCJ2E4MZJrIUrw9rCPus_ktCebK7_FYt3WOOpRF9BaH45dKNoFmbI7Vqw0Q0vkTsNu_KW7yFveuTItSTv5YYhumBKStz_ggVE9Z2XkZf7ohe-7WIQJo_rm3I7pNSu_jZ3B-j4AaWM3MDf41paeunAi_09KlMNqyv14u4i0rAFhIFzJT6xAdCHWizHHkgCw&token_type_hint=access_token, ctx=[ ssl_verify=false basic_auth_username=2pACQ7dLngWwxyT1PXbJ02ncFStZyD22 basic_auth_password=L8LjMTE8Jl8_LWsATN_GZoc8G6mnzL1ZG-XAQa4qXVhYvC7NrpShcyWuk9pywgQE hdr=[ Content-Type=application/x-www-form-urlencoded ] cookie=[ ] ]

**Reply by zandbelt** [reply]
it looks like you're still running an older version: you can see the version of liboauth2/mod_oauth2 printed in the logs at startup

**Reply by ravikaycee** [reply]
You are correct, old version is still used. I have followed these steps for installation,

RUN dpkg -i liboauth2_1.5.0-1.bullseye_amd64.deb
RUN dpkg -i liboauth2-apache_1.5.0-1.bullseye_amd64.deb
RUN dpkg -i libapache2-mod-oauth2_3.3.1-1.bullseye_amd64.deb
RUN dpkg -i libapache2-mod-auth-openidc_2.4.12.3-1.bullseye_amd64.deb

RUN sudo a2enmod auth_openidc
RUN sudo a2enmod oauth2

httpd.conf
LoadModule oauth2_module modules/mod_oauth2.so
LoadModule auth_openidc_module modules/mod_auth_openidc.so

httpd -k restart

Still the old version is used by apache
mod_oauth2-3.3.1-liboauth2-1.4.5.3 - init 

Could you please share some pointers to resolve this..

**Reply by ravikaycee** [reply]
Hi,
please let me know what needs to be corrected in the above mentioned steps to get the latest version of liboauth2. Is there any plans of releasing new version of mod_oauth2 with liboauth2 dep version >= 1.5.0

**Reply by zandbelt** [reply]
running `dpkg -l | grep "oauth2"` will tell you what is installed and what should be removed/added

also check your Dockerfile: you're using "sudo" for some commands, but not for the dpkg installation commands

**Reply by ravikaycee** [reply]
![image](https://user-images.githubusercontent.com/116255100/224493086-2c08f50c-31df-4c1b-80f1-753dfb75696e.png)


Results seems to be valid. I tried installing the file manually using sudo as well. Still it is referring the old version in the logs.

**Reply by zandbelt** [reply]
I just double-checked mine:
```
root@83ad1f97d2bb:/# dpkg -l | grep oauth2 | grep -v nginx
ii  libapache2-mod-oauth2          3.3.1-1.bullseye               amd64        OAuth 2.0 Resource Server module for Apache
ii  liboauth2                      1.5.0-1.bullseye               amd64        C library for OAuth 2.0 servers and clients
ii  liboauth2-apache               1.5.0-1.bullseye               amd64        Apache bindings for liboauth2
root@83ad1f97d2bb:/# cat /var/log/apache2/error.log | grep mod_oauth2 | tail -n 1
[Sat Mar 11 16:05:33.339424 2023] [:info] [pid 18:tid 139894854040896] mod_oauth2-3.3.1-liboauth2-1.5.0 - init
```
so it must be a local issue

**Reply by ravikaycee** [reply]
I have done apache setup with different image again, right from the scratch and ended up with the same lib version issue. Is it possible to share an updated mod_oauth2 with expected liboauth2 & liboauth2-apache dependency version as >= 1.5.0

**Reply by zandbelt** [reply]
that's not going to help; I just re-tried myself from scratch with the packages provided from the Github liboauth2 releases page and everything works as expected

**Reply by ravikaycee** [reply]
I understand. Let's check it out whether it helps or not. It will be really helpful if you could share the package with updated dependencies..
